README.md
# Chlu Query API

An API server that provides read access to Chlu data

WIP

CHANGELOG.md
# Chlu API Query Changelog

## v0.3.1

- updated deps

## v0.3.0

Breaking Changes:

- `/dids/:id/reviews/about` and `/dids/:id/reviews/writtenby` now return `{ rows, count }` datastructure

Other Changes:

- use chlu-ipfs-support 0.3
- fix starting without --postgres
- accept `limit` and `offset` query parameters in `/dids/:id/reviews/about` and `/dids/:id/reviews/writtenby`

## v0.2.0

- added CLI params for Chlu SQL DB

## v0.1.0

First release


tests/api.test.js
const ChluAPIQuery = require('../src')
const sinon = require('sinon')
const expect = require('chai').expect
const request = require('supertest')
const logger = require('chlu-ipfs-support/tests/utils/logger')

describe('HTTP server', () => {

    let chluApiQuery, chluIpfs, app
    let fakeReviewRecords, fakeDIDDocuments, fakeReviewsWrittenByDID, fakeReviewsAboutDID

    before(() => {
        // disable logs
        process.env.DISABLE_LOGS = '1'
    })

    beforeEach(() => {
        fakeReviewRecords = {
            'QmWBTzAwP8fz2zRsmzqUfSKEZ6GRTuPTsBVfJs6Y72D1hz': {
                review: 'hello world',
                editable: false
            }
        }
        fakeReviewsWrittenByDID = {
            'did:chlu:abc': [ { content: 'data' }]
        }
        fakeReviewsAboutDID = {
            'did:chlu:abc': [ { content: 'data' }]
        }
        fakeDIDDocuments = {
            'did:chlu:abc': {
                content: 'data'
            }
        }
        chluIpfs = {
            start: sinon.stub().resolves(),
            readReviewRecord: sinon.stub().callsFake(async x => {
                return fakeReviewRecords[x]
            }),
            getDID: sinon.stub().callsFake(async x => {
                return fakeDIDDocuments[x] || null
            }),
            getReviewsWrittenByDID: sinon.stub().callsFake(async x => {
                return fakeReviewsWrittenByDID[x] || []
            }),
            getReviewsAboutDID: sinon.stub().callsFake(async x => {
                return fakeReviewsAboutDID[x] || []
            }),
            logger: logger('API Server')
        }
        chluApiQuery = new ChluAPIQuery({ chluIpfs })
        sinon.spy(chluApiQuery, 'isDIDID')
        app = request(chluApiQuery.api)
    })

    it('starts correctly', async () => {
        expect(chluApiQuery.start).to.be.a('function')
        chluApiQuery.api = {
            listen: sinon.stub().callsFake((port, cb) => cb())
        }
        await chluApiQuery.start()
        expect(chluApiQuery.port).to.equal(3005)
        expect(chluIpfs.start.called).to.be.true
        expect(chluApiQuery.api.listen.calledWith(chluApiQuery.port)).to.be.true
    })

    it('/', async () => {
        await app.get('/').expect(200)
    })

    describe('/api/v1', () => {

        it('GET /reviews/:cid', async () => {
            await app.get('/api/v1/reviews/lol').expect(400)
            const multihash = 'QmWBTzAwP8fz2zRsmzqUfSKEZ6GRTuPTsBVfJs6Y72D1hz' // valid but not real
            await app.get(`/api/v1/reviews/${multihash}`)
                .expect('Content-Type', 'application/json; charset=utf-8')
                .expect(200, { review: 'hello world' })
            let options = chluIpfs.readReviewRecord.args[0][1]
            expect(options).to.deep.equal({
                validate: { throwErrors: false },
                getLatestVersion: true
            })
            await app.get(`/api/v1/reviews/${multihash}?getLatestVersion=false`)
                .expect('Content-Type', 'application/json; charset=utf-8')
                .expect(200, { review: 'hello world' })
            options = chluIpfs.readReviewRecord.args[1][1]
            expect(options).to.deep.equal({
                validate: { throwErrors: false },
                getLatestVersion: false
            })
        })

        it('GET /dids/:id', async () => {
            await app.get('/api/v1/dids/lol').expect(400)
            await app.get('/api/v1/dids/did:chlu:notexists').expect(404)
            await app.get('/api/v1/dids/did:chlu:abc')
                .expect(200, { content: 'data' })
            expect(chluIpfs.getDID.calledWith('did:chlu:abc')).to.be.true
            expect(chluApiQuery.isDIDID.calledWith('did:chlu:abc')).to.be.true
        })

        it('GET /dids/:id/reviews/writtenby', async () => {
            await app.get('/api/v1/dids/lol/reviews/writtenby').expect(400)
            await app.get('/api/v1/dids/did:chlu:abc/reviews/writtenby')
                .expect(200, [{ content: 'data' }])
            await app.get('/api/v1/dids/did:chlu:def/reviews/writtenby').expect(200, [])
            expect(chluApiQuery.isDIDID.calledWith('did:chlu:abc')).to.be.true
        })

        it('GET /dids/:id/reviews/about', async () => {
            await app.get('/api/v1/dids/lol/reviews/about').expect(400)
            await app.get('/api/v1/dids/did:chlu:abc/reviews/about')
                .expect(200, [{ content: 'data' }])
            await app.get('/api/v1/dids/did:chlu:def/reviews/about').expect(200, [])
            expect(chluApiQuery.isDIDID.calledWith('did:chlu:abc')).to.be.true
            expect(chluApiQuery.isDIDID.calledWith('did:chlu:def')).to.be.true
        })

    })

})

src/bin.js
#!/usr/bin/env node
const ChluAPIQuery = require('./index.js');
const ChluSQLIndex = require('chlu-ipfs-support/src/modules/orbitdb/indexes/sql')
const cli = require('commander');
const package = require('../package.json');

let server = null

function handleErrors(fn) {
    return function (...args) {
        fn(...args).catch(err => {
            console.log(err);
            console.trace(err);
            process.exit(1);
        });
    };
}

async function start(options){
    console.log('Starting Chlu API Query');
    if (!options.btc) {
        console.warn('\nWARNING: BTC Blockchain access through BlockCypher is strongly suggested\n');
    }
    const config = {
        network: options.network,
        directory: options.directory,
        blockCypherApiKey: options.btc,
        bitcoinNetwork: options.btcNetwork,
        OrbitDBIndex: ChluSQLIndex,
        OrbitDBIndexOptions: {
            dialect: options.postgres ? 'postgres' : undefined,
            database: options.databaseDb,
            host: options.databaseHost,
            port: options.databasePort,
            username: options.databaseUser,
            password: options.databasePassword,
            enableWrites: options.write,
            enableValidations: options.write
        },
    };
    server = new ChluAPIQuery({
        port: options.port,
        chluIpfsConfig: config
    });
    await server.start();
}

process.on('SIGINT', async function() {
    try {
        console.log('Stopping gracefully');
        if (server) await server.stop();
        console.log('Goodbye!');
        process.exit(0);
    } catch(exception) {
        console.trace(exception);
        process.exit(1);
    }
});


cli
    .name('chlu-api-query')
    .description('Reference implementation of the Chlu Query API. http://chlu.io')
    .version(package.version);

cli
    .command('start')
    .description('run the Chlu Query API Server')
    // Chlu specific options
    .option('-n, --network <network>', 'use a custom Chlu network instead of experimental')
    .option('-d, --directory <path>', 'where to store Chlu data, defaults to ~/.chlu-query')
    // Blockchain
    .option('--btc <token>', 'turn on BTC Blockchain access using a Blockcypher API Token. Other systems will be supported in the future')
    .option('--btc-network <network>', 'choose the BTC network you want to connect to. Default is test3')
    // ChluDB Options
    .option('--postgres', 'connect to PostgreSQL database, uses SQLite otherwise')
    .option('--no-write', 'disable database writes, useful if you have a Chlu Collector writing to the same database')
    .option('--database-host <s>', 'defaults to localhost')
    .option('--database-port <s>', 'defatuls to the default postgresql port')
    .option('--database-db <s>', 'defaults to \'chlu\'')
    .option('--database-user <s>', 'defatuls to \'chlu\'')
    .option('--database-password <s>', 'defatuls to empty password', null, '')
    .action(handleErrors(async cmd => {
        await start(cmd);
    }));

cli.parse(process.argv);

if (!process.argv.slice(2).length) {
    cli.help();
}

src/index.js
const path = require('path')
const { get } = require('lodash')
const ChluIPFS = require('chlu-ipfs-support')
const express = require('express')
const cors = require('cors')
const { isValidMultihash } = require('chlu-ipfs-support/src/utils/ipfs')

class ChluAPIQuery {
    constructor(config = {}) {
        const chluIpfsConfig = get(config, 'chluIpfsConfig', {})
        const directory = path.join(process.env.HOME, '.chlu-query')
        if (!chluIpfsConfig.directory) chluIpfsConfig.directory = directory
        this.chluIpfs = get(config, 'chluIpfs', null)
        if (!this.chluIpfs) this.chluIpfs = new ChluIPFS(chluIpfsConfig)
        this.port = get(config, 'port', 3005)
        this.logger = get(config, 'logger', this.chluIpfs.logger)
        this.prepareAPI()
        this.log = msg => this.logger.debug(`[API] ${msg}`)
    }

    async start() {
        await this.chluIpfs.start()
        this.log('Starting HTTP Server')
        await new Promise(resolve => this.api.listen(this.port, resolve))
        this.log(`Started HTTP Server on port ${this.port}`)
    }

    async stop() {
        await this.chluIpfs.stop()
    }

    isDIDID(didId) {
        return typeof didId === 'string' && didId.indexOf('did:') === 0
    }

    prepareAPI() {
        this.api = express()
        this.api.use(cors())
        this.api.get('/', (req, res) => res.send('Chlu API Query').end())
        const apiv1 = this.prepareAPIV1()
        this.api.use('/api/v1', apiv1)
    }

    prepareAPIV1() {
        const api = express()

        api.get('/reviews/:cid', async (req, res) => {
            const cid = req.params.cid
            const getLatestVersion = get(req, 'query.getLatestVersion', true) === 'false' ? false : true
            this.log(`Requested Review Record ${cid}, getLatestVersion: ${getLatestVersion ? 'yes' : 'no'}`)
            if (!isValidMultihash(cid)) {
                res.status(400).json(createError(`Multihash ${cid} is invalid`))
            } else {
                try {
                    const result = await this.chluIpfs.readReviewRecord(cid, {
                        getLatestVersion,
                        validate: {
                            throwErrors: false
                        }
                    })
                    if (typeof result === 'object') {
                        delete result.editable // that information is specific to the user's DID, so the client will have to recompute it
                    }
                    this.log(`Review Record ${cid}, getLatestVersion: ${getLatestVersion ? 'yes' : 'no'} => OK ${JSON.stringify(result)}`)
                    res.json(result)
                } catch (error) {
                    this.log(`Review Record ${cid}, getLatestVersion: ${getLatestVersion ? 'yes' : 'no'} => ERROR ${error.message}`)
                    console.error(error)
                    res.status(500).json(createError(error.message || 'Unknown Error'))
                }
            }
        })

        api.get('/dids/:id', async (req, res) => {
            const didId = req.params.id
            const waitUntilPresent = get(req, 'query.waitUntilPresent', false) === 'true'
            this.log(`Requested DID ${didId}, waitUntilPresent: ${waitUntilPresent ? 'yes' : 'no'}`)
            if (!this.isDIDID(didId)) {
                res.status(400).json(createError(`DID ID ${didId} is invalid`))
            } else {
                try {
                    const result = await this.chluIpfs.getDID(didId, waitUntilPresent)
                    if (result) {
                        this.log(`DID ${didId} => OK ${JSON.stringify(result)}`)
                        res.json(result)
                    } else {
                        // Not found
                        this.log(`DID ${didId} => NOT FOUND`)
                        res.status(404).json(createError(`DID ${didId} not found`))
                    }
                } catch (error) {
                    this.log(`DID ${didId} => ERROR ${error.message}`)
                    console.error(error)
                    res.status(500).json(createError(error.message || 'Unknown Error'))
                }
            }
        })

        api.get('/dids/:id/reviews/about', async (req, res) => {
            const didId = req.params.id
            const offset = req.query.offset
            const limit = req.query.limit
            this.log(`Requested Reviews about DID ${didId}: limit ${limit} offset ${offset}`)
            if (!this.isDIDID(didId)) {
                res.status(400).json(createError(`DID ID ${didId} is invalid`))
            } else {
                try {
                    const result = await this.chluIpfs.getReviewsAboutDID(didId, offset, limit)
                    if (result) {
                        this.log(`Reviews by DID ${didId} => OK ${JSON.stringify(result)}`)
                        res.json(result)
                    }
                } catch (error) {
                    this.log(`Reviews by DID ${didId} => ERROR ${error.message}`)
                    console.error(error)
                    res.status(500).json(createError(error.message || 'Unknown Error'))
                }
            }
        })

        api.get('/dids/:id/reviews/writtenby', async (req, res) => {
            const didId = req.params.id
            const offset = req.query.offset
            const limit = req.query.limit
            this.log(`Requested Reviews written by DID ${didId}`)
            if (!this.isDIDID(didId)) {
                res.status(400).json(createError(`DID ID ${didId} is invalid`))
            } else {
                try {
                    const result = await this.chluIpfs.getReviewsWrittenByDID(didId, offset, limit)
                    if (result) {
                        this.log(`Reviews written by DID ${didId} => OK ${JSON.stringify(result)}`)
                        res.json(result)
                    }
                } catch (error) {
                    this.log(`Reviews written by DID ${didId} => ERROR ${error.message}`)
                    console.error(error)
                    res.status(500).json(createError(error.message || 'Unknown Error'))
                }
            }
        })

        return api
    }

}

function createError(message) {
    return { message }
}

module.exports = ChluAPIQuery